drivers/hmis_external_apis/app/models/hmis_external_apis/external_forms/form_submission.rb

```diff
@@ -50,37 +53,65 @@ def self.from_raw_data(raw_data, object_key:, last_modified:, form_definition:,
       'json',
     ].map { |v| [v, "value_#{v}"] }
 
-    # note, we need to set all fields- bulk insert becomes unhappy if the columns are not uniform
-    def self.cde_value_fields(definition, value)
-      result = {}
-      VALUE_FIELDS.map do |field_type, field_name|
-        result[field_name] = field_type == definition.field_type ? value : nil
-      end
-      result
+    # keys that are added to raw_values upon submission, but are not part of the form values
+    EXTRANEOUS_KEYS = ['captcha_score', 'form_definition_id', 'form_content_digest'].freeze
+
+    def form_values
+      raw_data.reject { |key, _| EXTRANEOUS_KEYS.include?(key) }
     end
 
-    def process_custom_data_elements!(form_definition:)
-      custom_data_elements.delete_all
-      cdes = []
-      now = Time.current
-      form_definition.custom_data_element_definitions.each do |cded|
-        value = raw_data[cded.key]
-        next if value.blank?
-
-        cdes << {
-          owner_type: self.class.sti_name,
-          owner_id: id,
-          data_source_id: cded.data_source_id,
-          data_element_definition_id: cded.id,
-          UserID: cded.user_id,
-          DateCreated: now,
-          DateUpdated: now,
-        }.merge(HmisExternalApis::ExternalForms::FormSubmission.cde_value_fields(cded, value))
+    def run_form_processor(project, current_user)
+      values_to_process = form_values.clone
+
+      # Only if there are Client and/or Enrollment fields in the form definition, initialize an enrollment
+      # (which will in turn initialize a Client, inside the form processor).
+      if definition.creates_client_or_enrollment?
+        household_id = form_values['Enrollment.householdId']
```

Comment:
This logic looks good to me. The nested if-else is a little hard to follow, though, and we should also add some validation for the incoming values. Suggestion (untested!):

```ruby
household_id = form_values['Enrollment.householdId']
# drop household id if it doesn't match expected format
household_id = nil unless household_id.starts_with?('HH') && household_id.size > 20
# drop household id if it has a conflict in this data source in another project (just to be safe)
household_id = nil if Hmis::Hud::Enrollment.where(household_id: household_id, data_source: project.data_source).where.not(project: project).exists?
# if no household id, generate a new one
household_id ||= Hmis::Hud::Enrollment.generate_household_id

relationship_to_hoh = form_values['Enrollment.relationshipToHoH']
# todo: drop if form value for relationshipToHoH was an invalid value
# set default relationship_to_hoh (1 for new household, 99 for existing household)
relationship_to_hoh ||= project.enrollments.where(household_id: household_id).exists? ? 99 : 1

Reply:
Thanks, I like the restructure. I expanded on it a little bit to ensure that invalid household ID also nulls out the relationship. Also, in both cases, the values need to be dropped from the hash passed to the form processor, otherwise the form processor will override what we pass when creating the enrollment.